src/libhost1x/host1x-gr2d.c

```diff
@@ -269,19 +299,130 @@ int host1x_gr2d_clear(struct host1x_gr2d *gr2d,
 	return 0;
 }
 
+/*
+ * Returned values:
+ * 	0 = not intersected
+ *
+ * 	1 = intersected and source pixbuf area starts lower in the memory than
+ * 	    the destination area
+ *
+ * 	2 = intersected and source pixbuf area starts higher in the memory than
+ * 	    the destination area
+ */
+static int pixbuf_areas_intersect(struct host1x_pixelbuffer *src,
+				  struct host1x_pixelbuffer *dst,
+				  unsigned int sx, unsigned int sy,
+				  unsigned int dx, unsigned int dy,
+				  unsigned int width, unsigned int height)
+{
+	struct host1x_bo *src_orig = src->bo->wrapped ?: src->bo;
+	struct host1x_bo *dst_orig = dst->bo->wrapped ?: dst->bo;
+	uint32_t p0_top_left, p0_bottom_right, p0_pitch, p0_bpp;
+	uint32_t p1_top_left, p1_bottom_right, p1_pitch, p1_bpp;
+	uint32_t tmp;
+	int ret = 2;
+
+	/*
+	 * Areas are *not* intersected if their BO isn't common.
+	 */
+	if (src_orig != dst_orig)
+		return 0;
+
+	/*
+	 * p0 for the destination area
+	 * p1 for the source area
+	 */
+	p0_bpp          = PIX_BUF_FORMAT_BYTES(dst->format);
+	p0_pitch        = dst->pitch;
+	p0_top_left     = dst->bo->offset + p0_pitch * dy + dx * p0_bpp;
+	p0_bottom_right = p0_top_left + p0_pitch * height;
+
+	p1_bpp          = PIX_BUF_FORMAT_BYTES(src->format);
+	p1_pitch        = src->pitch;
+	p1_top_left     = src->bo->offset + p1_pitch * sy + sx * p1_bpp;
+	p1_bottom_right = p1_top_left + p1_pitch * height;
```

Comment:
I think this function is needlessly complicated, because it seems to allow interpreting overlapping memory as wildly different things. Sure, sub-allocations will allow multiple allocations of different format/pitches inside the same BO. But those allocations will not overlap. There's some use-cases where untyped textures are used (but not on T20-class hardware, more like OpenGL 4.x hardware), but in these cases the texel-size is known (you can overlay float32 on int32 and rgba8888 etc, but not on int16 etc). Two separate sub-allocations shouldn't overlap in memory.

So, I think we can first early-out on `src_orig != dst_orig`, then on `src_orig->offset != dst_orig->offset` (different sub-allocations). Then we can assert that the sizes are the same (just in case we somehow botched a sub-allocation). After that, the rest becomes a lot easier, as we *know* that the format and pitch have to be the same. Basially, we need to do a rectangle overlap test, and finally compare the top-left points.